### PR TITLE
fix(vrl): `parse_url` - treat repeated query strings as arrays

### LINKED ISSUE #376: parse_url doesn't support duplicated query parameters
### A note for the community

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Problem

For URLs like this one `http://exam.co/?key=val1&key=val2` when we use `parse_url` function, only the last value will be returned. 
Having duplicated query parameters is a valid HTTP URL
Screenshot 2022-02-25 at 11 25 11

This happens because the list of query params is turned in to a map [here](https://github.com/vectordotdev/vector/blob/26b4e959613dd1c6768f11ba2e1ff6e8df0494bd/lib/vrl/stdlib/src/parse_url.rs#L146)

### Configuration

```text
not needed as this is pure VRL.
```

### Version

vector 0.19.2

### Debug Output

_No response_

### Example Data

_No response_

### Additional Context

_No response_

### References

_No response_

### LINKED ISSUE #11576: parse_url doesn't support duplicated query parameters
### A note for the community

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Problem

For URLs like this one `http://exam.co/?key=val1&key=val2` when we use `parse_url` function, only the last value will be returned. 
Having duplicated query parameters is a valid HTTP URL
Screenshot 2022-02-25 at 11 25 11

This happens because the list of query params is turned in to a map [here](https://github.com/vectordotdev/vector/blob/26b4e959613dd1c6768f11ba2e1ff6e8df0494bd/lib/vrl/stdlib/src/parse_url.rs#L146)

### Configuration

```text
not needed as this is pure VRL.
```

### Version

vector 0.19.2

### Debug Output

_No response_

### Example Data

_No response_

### Additional Context

_No response_

### References

_No response_

### PR DESCRIPTION
**Pull request by leebenson**
Small PR to align the handling of `parse_url` query strings with `parse_query_string`.

Closes vectordotdev/vrl#376

Signed-off-by: Lee Benson <lee@leebenson.com>

### CODE REVIEW

**Review by spencergilbert — APPROVED** [review]

**Review by jszwedko — COMMENTED** [review]
Thanks for adding the release highlight comment!

I think we may need to update `inner_kind` here too since the type definition of `query` is now different? I think you can borrow the type def from `parse_query_string`.

**Comment by leebenson** [comment]
Would like to DRY this up a bit. Seems daft to have the same logic repeated for `parse_url` and `parse_query_string`. Will put it into draft for now.

File: lib/vrl/stdlib/src/parse_url.rs
Comment (leebenson) [inline lib/vrl/stdlib/src/parse_url.rs:69]: This is exactly the same code as before, just rejigged by CLion to match the trait definition order.

**Review by spencergilbert — APPROVED** [review]
👍 (aside from CI 🔴 )

**Review by jszwedko — APPROVED** [review]
Looks good! Thanks for refactoring this to share the implementation.

**Comment by jszwedko** [comment]
> Hello! I was just thinking that this change will potentially break a number of our feed processing because we use fixed schemas where `query` is strictly a map of strings.
> 
> To preserve the old behaviour, we will need to do something like this:
> 
> ```coffee
> parts = parse_url!(some_url)
> parts.query = if is_array(parts.query) { parts.query[-1] } else { parts.query }
> ```
> 
> Perhaps this snippet would be good to mention in the breaking changes section for people who needs to preserve the old behaviour ?

Thanks for the note!

I'm not sure I follow though. `parts.query` will continue to be a map with string keys. The difference is that now the value for one of those keys could potentially be an array if the query string had the same key repeated twice (like `?foo=1&foo=2`). Would that still cause compatibility issues for you?

**Comment by hhromic** [comment]
> I'm not sure I follow though. `parts.query` will continue to be a map with string keys. The difference is that now the value for one of those keys could potentially be an array if the query string had the same key repeated twice (like `?foo=1&foo=2`). Would that still cause compatibility issues for you?

Ah sorry, I made the example wrong. Indeed `query` is a map of string keys, but now the values can be strings or arrays, whereas before they could only be strings.

I just realised that without iteration support, then there is no way to preserve the old behaviour, which means the breaking change is a bigger problem than I was anticipating :(

Would you consider adding an argument to `parse_url()` so it preserves the old behaviour? For us it doesn't have to be the default. For example something like `single_query_keys=true|false`.

**Comment by jszwedko** [comment]
> > I'm not sure I follow though. `parts.query` will continue to be a map with string keys. The difference is that now the value for one of those keys could potentially be an array if the query string had the same key repeated twice (like `?foo=1&foo=2`). Would that still cause compatibility issues for you?
> 
> Ah sorry, I made the example wrong. Indeed `query` is a map of string keys, but now the values can be strings or arrays, whereas before they could only be strings.
> 
> I just realised that without iteration support, then there is no way to preserve the old behaviour, which means the breaking change is a bigger problem than I was anticipating :(
> 
> Would you consider adding an argument to `parse_url()` so it preserves the old behaviour? For us it doesn't have to be the default. For example something like `single_query_keys=true|false`.

👍 that seems reasonable until iteration support lands.

You mentioned having a fixed schema though, would you ever actually have duplicate keys in query strings?

**Comment by jszwedko** [comment]
Also, thanks for catching this on the PR before we had to have this conversation following a release 😄

**Comment by hhromic** [comment]
> You mentioned having a fixed schema though, would you ever actually have duplicate keys in query strings?

Unfortunately we don't control what the query strings will contain. In our case, we are processing webproxy logs, therefore the query strings can be virtually anything. I pretty much expect most use-cases of `parse_url()` to share this assumption :(

**Review by jszwedko — CHANGES_REQUESTED** [review]
Just marking this as "needs changes" to avoid an automerge until we address @hhromic's concerns.

**Comment by leebenson** [comment]
I'm happy to add a flag to both `parse_url` and `parse_query_string` to toggle between the two modes.

Maybe `query_string_use_last_value`, defaulting to false?

@jszwedko is good with names -- I'll defer to his approval 😄

**Comment by jszwedko** [comment]
Thinking about this more, I think it's a bit odd to have mixed types for the values depending on whether they appear more than once (`[]string`) or not (`string). I think I'm inclined to have it _always_ return a map of `string` -> `[]string`. I think this would make using the map simpler, given VRL's type handling, since all of the values would be of the same type.

We can still add a flag to vary this behavior until iteration support is added, though, as dealing with arrays of values is otherwise not easy.

Maybe something like:

* `parse_url(url, multiple_query_values: first|last|all)`

? If `all`, `values` is always a `[]string`, otherwise it takes the `first` or `last` one depending?

Alternatively, we don't return a map, but instead an array of "tuples", but that'd be even harder to work with in VRL right now.

**Comment by leebenson** [comment]
👍🏻 I'm a fan of that API, @jszwedko, but I'd propose the following tweaks:

```
parse_url(url, query_values: default|first|last|array)
```

The new values being:

- `default`: The current behaviour of _parse_query_string_. Single values are strings; multi are arrays.
- `array`: The same as your 'all' option, but clearer alongside a rename to _query_values_ that every value is an array.

I think the issue is that we're effectively trying to mimic server-like HTTP request parsing, without language support that would make _using_ it simple.

This would give us an opportunity to align the behaviour of _parse_url_ with _parse_query_string_, without changing the signature for both.

And once we have iteration support, the default behaviour (which I'd expect as a user) is easier to work with in a dynamic context.

What do you think?

**Comment by hhromic** [comment]
> 👍🏻 I'm a fan of that API, @jszwedko, but I'd propose the following tweaks:
> 
> ```
> parse_url(url, query_values: default|first|last|array)
> ```
> 
> The new values being:
> 
> * `default`: The current behaviour of _parse_query_string_. Single values are strings; multi are arrays.
> * `array`: The same as your 'all' option, but clearer alongside a rename to _query_values_ that every value is an array.
> 

I really like the API shown here 👍 
Minor nitpick: perhaps the `default` setting could be named `mixed` instead given that it returns strings or arrays of strings.

> I think the issue is that we're effectively trying to mimic server-like HTTP request parsing, without language support that would make _using_ it simple.

I agree with this. VRL in the end is about making common processing use-cases simple. I'm fairly sure that the _majority_ of Vector users do not need to process query keys with such precision. I bet many users are not even aware that this is even possible in the HTTP spec :)  Also I'm pretty sure that in the real-world, repeated query keys are very rare too.

> 
> This would give us an opportunity to align the behaviour of _parse_url_ with _parse_query_string_, without changing the signature for both.

Agreed too and I think is important that complementary function signatures are as similar as possible.

> 
> And once we have iteration support, the default behaviour (which I'd expect as a user) is easier to work with in a dynamic context.

I think that the `array` setting is the most friendly if we are going to do iteration over the results of the parsing. Having mixed-type values will require checking first if the result is an array before iterating (assuming that iteration support fails on non-iterables, or if iteration-support iterates strings character-by-character). Will depend on how iteration supports is implemented.

File: lib/vrl/stdlib/src/url_util.rs
Comment (StephenWakely) [inline lib/vrl/stdlib/src/url_util.rs:6]: I believe we know that if it is an array it is going to be an array of bytes, so this should be possible:

```suggestion
    Collection::from_unknown(Kind::bytes().or_array(Collection::from_unknown(Kind::bytes()))
```

**Comment by leebenson** [comment]
Moving this back into draft temporarily to implement the version with a `query_values` flag, as discussed above.

**Comment by jszwedko** [comment]
Closing this as stale since it seems unlikely to be picked up at this point.

**Comment by anas-aso** [comment]
@jszwedko is this something that is dropped once and for all ? It seems like it got dragged because of the missing consensus on how the breaking change should look like. I have a suggestion that might unblock users without any breaking change.
If a query parameter is repeated, just append the new value to the previous one(s) and make them comma separated (or any other separator of your choice).
```rust
$ v=s'https://examp.le/?key=val1&key=val2'
"https://examp.le/?key=val1&key=val2"

$ parsed, _ = parse_url(v)
{ "fragment": null, "host": "examp.le", "password": "", "path": "/", "port": null, "query": { "key": "val1,val2" }, "scheme": "https", "username": "" }

$ parsed.query
{ "key": "val1,val2" }

$ split(parsed.query.key, ",")
["val1", "val2"]
```

Having something like this would unblock anyone wanting to get the values for repeated query strings without breaking the usage or the type of the returned values type.

**Comment by hhromic** [comment]
@anas-aso , that is not a robust solution unfortunately.

Consider this URL: `https://examp.le/?key=val1,&key=,val2`
I know is a "weird looking" URL but it is still valid.

With your proposal, this would happen:
```coffee
$ v=s'https://examp.le/?key=val1,&key=,val2'
"https://examp.le/?key=val1,&key=,val2"

$ parsed, _ = parse_url(v)
{ "fragment": null, "host": "examp.le", "password": "", "path": "/", "port": null, "query": { "key": "val1,,,val2" }, "scheme": "https", "username": "" }

$ parsed.query
{ "key": "val1,,,val2" }

$ split(parsed.query.key, ",")
["val1", "", "", "val2"]
```

That's the reason we need to implement a proper multi-dict representation.

HTTPbin offers a good example:
```
$ curl 'http://httpbin.org/get?key=val1,&key=,val2'
{
  "args": {
    "key": [
      "val1,",
      ",val2"
    ]
  },
  "headers": {
    "Accept": "*/*",
    "Host": "httpbin.org",
    "User-Agent": "curl/7.88.1",
    "X-Amzn-Trace-Id": "Root=1-64d213e7-400dfe24602e81a94a19c219"
  },
  "origin": "81.193.174.181, 165.225.92.210",
  "url": "http://httpbin.org/get?key=val1,&key=,val2"
}
```

**Comment by hhromic** [comment]
@anas-aso , that is not a robust solution unfortunately.

Consider this URL: `https://examp.le/?key=val1,&key=,val2`
I know is a "weird looking" URL but it is still valid.

With your proposal, this would happen:
```coffee
$ v=s'https://examp.le/?key=val1,&key=,val2'
"https://examp.le/?key=val1,&key=,val2"

$ parsed, _ = parse_url(v)
{ "fragment": null, "host": "examp.le", "password": "", "path": "/", "port": null, "query": { "key": "val1,,,val2" }, "scheme": "https", "username": "" }

$ parsed.query
{ "key": "val1,,,val2" }

$ split(parsed.query.key, ",")
["val1", "", "", "val2"]
```

That's the reason we need to implement a proper multi-dict representation.

HTTPbin offers a good example:
```
$ curl 'http://httpbin.org/get?key=val1,&key=,val2'
{
  "args": {
    "key": [
      "val1,",
      ",val2"
    ]
  },
  // ...
}
```

**Comment by anas-aso** [comment]
Thanks @hhromic for the quick reply. I can understand why multi-dict is the correct solution, but it's a bit disappointing that this issue has been open for 18 months and the only PR trying to fix it was discarded. So I was hopping to move it forward even if it's a quick and dirty solution.

**Comment by hhromic** [comment]
Hey @anas-aso , I'm just a humble external contributor to Vector, so is not my call unfortunately.
In my opinion, the original issue you opened (#11576) should indeed be moved to the new VRL repository.
From my side, I would love this issue to be fixed/addressed too, as it is part of the HTTP spec.

**Comment by jszwedko** [comment]
Thanks for bumping this! I transferred [the issue](https://github.com/vectordotdev/vrl/issues/376) to the VRL repository.

I agree that it would be nice to fix. I think the simplest way to fix it, while maintaining compatibility, would be the extra flag that allows opting into multi-value keys. I'd be happy to see someone pick this up if they feel motivated.

**Pull request closed without merging**
